README.md

```diff
@@ -1,16 +1,1831 @@
+        }
+      }
+    }
+  },
+  "roots": [
+    {
+      "/": "bafyreicc3rilzaxravqju766yr4v7p6nm6sjsd4lbkfa4uqohbbxaejh7e"
+    }
+  ]
+}
+```
+
+# 6 Result
+
+A `Result` records the output of the [Task], as well as its success or failure state.
+
+## 6.1 Schema
+
+```ipldsch
+type Result union {
+  | any "ok"
+  | any "error"
+} representation keyed
+}
+```
+
+## 6.2 Variants
+
+## 6.2.1 Success
+
+The success branch MUST contain the value returned from a successful [Task] wrapped in the `"ok"` tag. The exact shape of the returned data is left undefined to allow for flexibility in various Task types.
+
+```json
+{ "ok": 42 }
+```
+
+## 6.2.2 Failure
+
+The failure branch MAY contain detail about why execution failed wrapped in the "error" tag. It is left undefined in this specification to allow for [Task] types to standardize the data that makes sense in their contexts.
+
+If no information is available, this field SHOULD be set to `{}`.
+
+```json
+{
+  "error": {
+    "dev/reason": "unauthorized",
+    "http/status": 401
+  }
+}
+```
+
+# 7 Effect
+
+Workflows often consist of many, sometimes concurrent, steps which form an execution threads. Steps of such workflows are discrete [Task] [Invocation]s, each producing some [Result] and a description of subsequent step(s) caused by it. Subsequent steps are denoted with an `Effect`, a set of [Invocation]s that [Executor] must perform to advance workflow execution.
+
+All of the [Invocation]s of the `Effect` MUST be concurrent with one another, unless explicitly arranged differently using [Pipelines].
+
+Effect MAY instruct the [Executor] to run set of [Task] [Invocation]s in concurrent execution threads by providing them as list under `fork` field.
+
+Effect MAY instruct the [Executor] that the [Task] [Invocation] is a continuation of the current execution thread by providing it under `join` field.
+
+Often [Invocation] in the `join` field will synthesize [Result]s of the concurrent execution threads (spawned by `fork`), and incorporate them into the current execution thread.
+
+It is also possible for the thread to complete execution (omit `join` field) and leave concurrent threads behind. The `join` field, simply marks the execution thread and provides a way to follow it through trail of [Receipt]s.
+
+## 7.1 Schema
+
+```ipldsch
+# Represents a request to invoke enclosed set of tasks concurrently
+type Effect {
+  # Primary set of tasks to be invoked
+  fork      [&Invocation]
+  
+  # Additional task to be invoked with added semantics
+  # of representing a workflow execution continuation.
+  join       optional &Invocation
+}
+```
+
+## 7.2 Fields
+
+### 7.2.1 Forked Task Invocations
+
+The OPTIONAL `fork` field, if present MUST be a list of an alphabetically ordered [Invocation] links. List MUST NOT not contain duplicate entries. Every linked [Invocation] MUST have `cause` field set to the same [Invocation] (link) as the `run` field of the containing [Receipt].
```

Comment:
Is the field optional, or can the array be empty? (Or both)? The schema above is `fork [&Invocation]`.

Reply:
That's surprising! TIL!